### DISCUSSION TITLE
Some question about implement customized history manager

### DISCUSSION THREAD

**Discussion by PuffMeow** [pmndrs/valtio#710] [Q&A]
The proxyWithHistory can't meet my need, so I want to implement a  customized history manager by myself. But I meet some question.
Can you give me some advice how to resolve this problem? @dai-shi 
What I do is like this:

```js
const store = proxy({
  myStates: {}
})

class History {
  snapshots: [];
  index: -1;
  save(state) {
   this.snapshots.push(state)
   ++this.index;
  }
  redo() {
   // When I call redo function, to re-render my components.
   // But this time it will fire the subscribe and save the snapshot again.
   setStore(store, this.snapshots[--this.index]);
  }
}

const history = new History();

// everytime store value changes, it will save the snapshot
subscribe(store, () => {
  history.save(store);
})
```

**Comment by dai-shi** [comment]
Yeah, we need some hacks there. You want to read `proxyWithHistory` code carefully.

> so I want to implement a customized history manager by myself.

I want to point out that creating a customized one is reasonable, because `proxyWithHisotry` doesn't cover all use cases.
